lib/ts/MemArena.cc

```diff
@@ -0,0 +1,232 @@
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+ */
+
+#include "MemArena.h"
+#include <ts/ink_memory.h>
+#include <ts/ink_assert.h>
+
+using namespace ts;
+
+inline void
+MemArena::Block::init(size_t n)
+{
+  size      = n;
+  allocated = 0;
+  next      = nullptr;
+}
+
+inline char *
+MemArena::Block::data()
+{
+  return reinterpret_cast<char *>(this + sizeof(Block));
+}
+
+/**
+    Allocates a new internal block of memory. If there are no existing blocks, this becomes the head of the
+     @allocations ll. If there are existing allocations, the new block is inserted in the allocations list.
+     If @back == true, the new block is appended and becomes the @current internal block.
+        @back == false, the new block is pushed into the generation but @current doesn't change.
+  */
+inline MemArena::Block *
+MemArena::newInternalBlock(size_t n, bool back)
+{
+  // Adjust to the nearest power of two. Works for 32 bit values which is ok as we are unlikely to
+  //  allocate over 4GB in this class. Allocate Block header and actual underlying memory together for locality. 
+  //  -16 to account for malloc/free headers.
+  static constexpr size_t free_space_per_page = DEFAULT_PAGE_SIZE - sizeof(Block) - 16;
+
+  void *tmp;
+  if( n <= free_space_per_page) { // will fit within one page, just allocate. 
+    tmp = ats_malloc(n + sizeof(Block));
+  } else {
+    size_t t = n;
+    t--;
+    t |= t >> 1;
+    t |= t >> 2;
+    t |= t >> 4;
+    t |= t >> 8;
+    t |= t >> 16;
+    t++;
+    n = t - sizeof(Block) - 16;
+    tmp = ats_malloc(t - 16);
+  }
+
+  auto block = std::shared_ptr<Block>(new (tmp) Block); // placement new
+  block->init(n);
+
+  if (allocations) {
+    arena_size += n;
+    generation_size += n;
+
+    if (back) { // append
+      current->next = block;
+      current       = current->next;
+      return current.get();
+    } else {
+      // Situation where we do not have enough space for a large block of memory. We don't want
+      //  to update @current because it would be wasting memory. Create a new block for the entire
+      //  allocation and just add it to the generation.
+      if (generation->next) {
+        std::shared_ptr<Block> tmp = generation->next;
+        generation->next           = block;
+        block->next                = tmp;
+      } else { // single block
+        block->next = generation;
+        generation = allocations = block;
+      }
+    }
+  } else { // empty
+    generation_size = n;
+    arena_size      = n;
+
+    allocations = block;
+    generation = current = allocations;
+  }
+
+  return block.get();
+}
+
+MemArena::MemArena()
+{
+  newInternalBlock(arena_size, 0); // nDefault size
+}
+
+MemArena::MemArena(size_t n)
+{
+  newInternalBlock(n, 0);
+}
+
+MemArena::~MemArena()
+{
+  empty();
```

Comment:
I think this is unnecessary. The `shared_ptr` instances will clear themselves on destruct, which should trigger cleaning up all the memory. Clearing the other member is pointless in the destructor.

Reply:
Removed.